File: additional exercise series/series_06 [Python]/Chromosomal crossover.py
def crossoverpoints(sequence1: list, sequence2: list):
    return len(_crossoverpoints(sequence1, sequence2))


def _crossoverpoints(sequence1: list, sequence2: list):
    return set(sequence1) & set(sequence2)


def chromosomal_partial_sum(sequence: list, crosses):
    sum, seq = 0, list()
    for item in sequence:
        if item in crosses:
            seq.append(sum)
            seq.append(item)
            sum = 0
        else:
            sum += item
    seq.append(sum)
    return seq


def maximalSum(sequence1: list, sequence2: list):
    crosses = _crossoverpoints(sequence1, sequence2)
    d1 = chromosomal_partial_sum(sequence1, crosses)
    d2 = chromosomal_partial_sum(sequence2, crosses)
    return sum(list(map(max, zip(d1, d2))))

File: exercise series (2019–2020)/week_08 [JavaScript]/Chromosomal crossover.js
function crossoverpunten(sequence1, sequence2) {
    return _crossoverpunten(sequence1, sequence2).length;
}

function _crossoverpunten(sequence1, sequence2) {
    return sequence1.filter(value => sequence2.includes(value));
}

function chromosomale_deelsom(sequence, crosses) {
    let sum = 0;
    let seq = [];
    for (let value of sequence) {
        if (crosses.includes(value)) {
            seq.push(sum);
            seq.push(value);
            sum = 0;
        } else {
            sum += value;
        }
    }
    seq.push(sum);
    return seq;
}

function maximaleSom(sequence1, sequence2) {
    const crosses = _crossoverpunten(sequence1, sequence2);
    const [d1, d2] = [chromosomale_deelsom(sequence1, crosses), chromosomale_deelsom(sequence2, crosses)];
    return d1.map((value, index) => Math.max(value, d2[index])).reduce((x, y) => x + y, 0);
}

chromosome1 = [3, 5, 7, 9, 20, 25, 30, 40, 55, 56, 57, 60, 62];
chromosome2 = [1, 4, 7, 11, 14, 25, 44, 47, 55, 57, 100];
console.log(_crossoverpunten(chromosome1, chromosome2))
console.log(chromosomale_deelsom(chromosome1, _crossoverpunten(chromosome1, chromosome2)));
console.log(crossoverpunten(chromosome1, chromosome2));
console.log(maximaleSom(chromosome1, chromosome2));